lib/frameworks/jasmine.py

```diff
@@ -1,16 +1,187 @@
-import json
+import json, retrying, urllib
+from selenium.webdriver.support.ui import WebDriverWait
+from itertools import groupby
+from xml.etree import ElementTree
+from lib import log
+
+
+def runTests( driver, url, timeout ):
+
+    # timeout is ignored
+
+    log.write( "Calculating number of tests ... " )
+    tests = getTests( driver, url )
+
+    log.writeln( "%d tests found" % len( tests ) )
+
+    testResults = []
+    counter = 0
+
+    for test in tests:
+
+        retries = 5
+        passed = False
+        testResult = None
+        exceptions = []
+        counter += 1
+
+        log.write( "Running test %d: %s ... " % ( counter, test ) )
+
+        while not passed and retries:
+
+            retries -= 1
+
+            try:
+
+                testResult = runTest( driver, "%s?spec=%s" % ( url, urllib.quote( test ) ) )
+
+                if isPassed( testResult[ "json" ] ):
+                    passed = True
+
+            except Exception as exception:
+
+                exceptions.append( exception )
+
+        if testResult:
+
+            log.writeln( "passed" ) if passed else log.writeln( "failed" )
+            testResults.append( testResult )
+
+        else:
+
+            log.writeln( "error" )
+            log.writeln( "Test skipped: \"%s\", exceptions: [%s]" % ( test, "\n".join( [ str( exception ) for exception in exceptions ] ) ) )
```

Comment:
We shouldn't let suite to be green if we are unable to gather any test results.
I believe preferable solution is to throw exception and stop execution.